experimental/oak_baremetal_channel/src/lib.rs

```diff
@@ -350,28 +50,70 @@ where
 {
     pub fn new(socket: T) -> Self {
         Self {
-            inner: Framed::new(socket),
+            inner: frame::Framed::new(socket),
         }
     }
-    fn recursively_complete_partial_message(
-        &mut self,
-        partial_message: PartialMessage,
-    ) -> anyhow::Result<Message> {
-        let frame = self.inner.read_frame()?;
-        match partial_message.add_frame(frame)? {
-            CompletionResult::Complete(message) => Ok(message),
-            CompletionResult::Incomplete(partial_message) => {
-                self.recursively_complete_partial_message(partial_message)
+    pub fn read_message<M: message::Message>(&mut self) -> anyhow::Result<M> {
+        let mut encoded_message: Vec<u8> = Vec::new();
+        let mut first_frame = self.inner.read_frame().context("failed to read frame.")?;
+
+        match (
```

Comment:
nit: this is just personal preference, but I think splitting this into two distinct if statements will make the code easier to read:

```
if !first_frame.flags.contains(frame::Flags::START)  {
    anyhow::bail!(...);
}
```

And then a separate if statemoer on whether the end bit is set or not.